Vialer/VoIPGRIDRequestOperationManager.m

```diff
@@ -305,12 +291,65 @@ - (void)setHandleAuthorizationRedirectForRequest:(NSURLRequest *)request andOper
+            NSString *appAccountUrl = [responseObject objectForKey:@"app_account"];
+            [self fetchSipAccountFromAppAccountURL:appAccountUrl withSuccess:^(NSString *sipUsername, NSString *sipPassword) {
+                [self setSipAccount:sipUsername andSipPassword:sipPassword];
+                if (success) success ();
+            } failure:^(NSError *error) {
+                if (failure) failure(error);
+            }];
+        } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
+            if (failure) failure(error);
+        }];
+        
+        [self setHandleAuthorizationRedirectForRequest:request andOperation:operation];
+        [self.operationQueue addOperation:operation];
+    }
+}
+
+/**
+ * Given an URL to an App specific SIP account (phoneaccount) this function fetches and sets the SIP account details for use in the app if any, otherwise the SIP data is set to nil.
+ * @param appAccountURL the URL from where to fetch the SIP account details e.g. /api/phoneaccount/basic/phoneaccount/XXXXX/
+ */
+- (void)fetchSipAccountFromAppAccountURL:(NSString *)appAccountURL withSuccess:(void (^)(NSString *sipUsername, NSString *sipPassword))success failure:(void (^)(NSError *error))failure {
+    if ([appAccountURL isKindOfClass:[NSString class]])
+        [self retrievePhoneAccountForUrl:appAccountURL success:^(AFHTTPRequestOperation *operation, id responseObject) {
+            NSObject *account = [responseObject objectForKey:@"account_id"];
+            NSObject *password = [responseObject objectForKey:@"password"];
+            if ([account isKindOfClass:[NSNumber class]] && [password isKindOfClass:[NSString class]]) {
+                if (success) success([(NSNumber *)account stringValue], (NSString *)password);
+            } else {
+                //No information about SIP account found, removed by user
+                if (success) success(nil, nil);
+            }
+        } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
+            if (failure) failure(error);
+        }];
+    else
+        //No URL supplied for Mobile app account, this means the user does not have the account set.
+        //We are also going to unset it by calling success with but sipUsername and sipPassword set to nil
+        if (success) success(nil, nil);
+}
+
 - (void)setSipAccount:(NSString *)sipAccount andSipPassword:(NSString *)sipPassword {
     if (sipAccount) {
         [[NSUserDefaults standardUserDefaults] setObject:sipAccount forKey:@"SIPAccount"];
         [SSKeychain setPassword:sipPassword forService:[[self class] serviceName] account:sipAccount];
         NSLog(@"Setting SIP Account %@", sipAccount);
     } else {
+        //First unregister the account with the middleware
+        [[PZPushMiddleware sharedInstance] unregisterSipAccount:self.sipAccount];
```

Comment:
This is the second I place I see it being used. My question remains: why should this be a singleton? You could request the property from the AppDelegate class.